tutorials/15-external-samplers/external-samplers.jmd

```diff
@@ -0,0 +1,168 @@
+nuts = AdvancedHMC.NUTS(n_adapts, tap; metric=metric)
+
+# Sample
+chain = sample(model, externalsampler(nuts), 10_000)
+```
+
+## Using new inference methods
+
+So far we have used to Turing's support for external samplers to go beyond the capabilities of the wrappers.
+Now we want to use this support to employ a sampler not supported within Turing's ecosystem yet.
+We will use the recently developped Micro-Cannoncial Hamiltonian Monte Carlo (MCHMC) sampler to showcase this.
+MCHMC[^3,^4] ((MCHMC's GitHub)[https://github.com/JaimeRZP/MicroCanonicalHMC.jl]) is HMC sampler that conserves energy along the exploration of the phase space.
+This is achieved by simulating the dynamics of a microcanonical Hamiltonian with an additional noise term to ensure ergodicity.
+
+Using this as well as other inference methods outside the Turing ecosystem is as simple as executing the code shown below:
+
+```julia
+using MicroCanonicalHMC
+# Create MCHMC sampler
+n_adapts = 1_000 # adaptation steps
+tev = 0.01 # target energy variance
+mchmc = MCHMC(n_adapts, tev; adaptive=true)
+
+# Sample
+chain = sample(model, externalsampler(mchmc), 10_000)
+```
+
+## The `AbstractMCMC` API
+
+As previously stated, in order to use external sampling libraries within `Turing` they must follow the `AbstractMCMC` API.
+In this section we will briefly dwell in what this entails.
+First and foremost, the inference method must be of the type `typeof(spl)<:AbstractMCMC.AbstractSampler`.
+Second, the stepping function of the MCMC algorithm must be made into instances of `AbstractMCMC.step` and follow the structure below:
+
+```
+# First step
+function AbstractMCMC.step{T<:AbstractMCMC.AbstractSampler}(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    spl::T;
+    kwargs...,
+)
+    [...]
+    return transition, sample
+end
+
+# N+1 step
+function AbstractMCMC.step{T<:AbstractMCMC.AbstractSampler}(
+    rng::Random.AbstractRNG,
+    model::AbstractMCMC.LogDensityModel,
+    sampler::T,
+    state;
+    kwargs...,
+) 
+    [...]
+    return transition, sample
+end
+```
+
+There are several characteristic to note in these functions:
+
+  - There must be two `step` functions:
+    
+      + A function that performs the first step which and initializes the sampler.
+      + A function that performs the following steps and which takes and extra input, `state`, which carries the initialization information.
+
+  - The functions must follow the displayed signatures.
+  - The output of the functions must be a transition, the current state of the sampler, and a sample, what is saved to the MCMC chain.
+
+The last requirement is that the transition must be structure with a field named `θ` which constains the values of the parameters of the model for said transition.
+This allows `Turing` to seemlessly extract the parameter values at each step of the chain when bundling the chains.
```

Comment:
This part is still something I'm not too keen on. Basically, the current way of converting a transition to something that bundle_samples can handle is really not great and I don't expect anyone outside of Turing.jl to implement this :confused:

Reply:
I think for the time being it is quite an important consideration for devs who want to make the connection as seemless as possible. 
We can change it as we change the code.

Reply:
For certain, but it's an aspect of the implementation that is bound to change very rapidly from the current state, to the point where I really don't want to have downstream packages rely on it (the result is that we might end up making non-breaking release, which subsequently ends up breaking downstream packages because we encouraged usage of internal and volatile code :confused: )

So I'd rather just say "Right now, it's still non-trivial to make it so the resulting samples are also bundled nicely into a `MCMCChains.Chains`, but this is being worked on (see this issue)." and link to some issue where we keep track of this:)